File: backend/types/sync.ts
import { z } from "npm:@hono/zod-openapi";
import {
Invite,
InviteCode,
Member,
	Message,
	MessageId,
	MessageVersionId,
	Role,
	RoleId,
	Room,
	RoomId,
	Session,
	SessionId,
	Thread,
	ThreadId,
	User,
	UserId,
} from "./common.ts";

export const MessageClient = z.union([
	z.object({
		type: z.literal("hello"),
		token: z.string(),
		last_id: z.string().optional(),
	}),
	z.object({ type: z.literal("pong") }),
]);

// const MessageRoom = z.union([
// ]);

export const MessageServer = z.union([
	z.object({ type: z.literal("ping") }),
	z.object({ type: z.literal("ready"), user: User }),
	z.object({ type: z.literal("error"), error: z.string() }),
	z.object({ type: z.literal("upsert.room"), room: Room }),
	z.object({ type: z.literal("upsert.thread"), thread: Thread }),
	z.object({ type: z.literal("upsert.message"), message: Message }),
	z.object({ type: z.literal("upsert.user"), user: User }),
	z.object({ type: z.literal("upsert.member"), member: Member }),
	z.object({ type: z.literal("upsert.session"), session: Session }),
	z.object({ type: z.literal("upsert.role"), role: Role }),
	z.object({ type: z.literal("upsert.invite"), invite: Invite }),
	// z.object({ type: z.literal("create.ban"), ban: Role }),
	z.object({ type: z.literal("delete.message"), id: MessageId }),
	z.object({ type: z.literal("delete.message_version"), id: MessageVersionId }),
	z.object({ type: z.literal("delete.user"), id: UserId }),
	z.object({ type: z.literal("delete.session"), id: SessionId }),
	z.object({ type: z.literal("delete.role"), id: RoleId }),
	z.object({ type: z.literal("delete.member"), room_id: RoomId, user_id: UserId }),
	z.object({ type: z.literal("delete.invite"), code: InviteCode }),
]);

/*
// expose a sse route per room..? would be nice if there was auth though
// return streamSSE(c, async (stream) => {
//   while (true) {
//     const message = `It is ${new Date().toISOString()}`
//     await stream.writeSSE({
//       data: message,
//       event: 'time-update',
//       id: String(id++),
//     })
//     await stream.sleep(1000)
//   }
// })
*/

File: crate-backend/src/types/sync.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use uuid::Uuid;

use super::{
    Invite, InviteCode, Message, MessageId, MessageVerId, Role, RoleId, Room, RoomId, RoomMember,
    Session, SessionId, Thread, ThreadId, User, UserId,
};

#[derive(Debug, PartialEq, Eq, ToSchema, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum MessageClient {
    Hello {
        token: String,

        // TODO: resutming connections,
        last_id: Option<String>,
    },
    Pong,
}

#[derive(Debug, Clone, PartialEq, Eq, ToSchema, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum MessageServer {
    Ping {},
    Ready {
        user: User,
    },
    Error {
        error: String,
    },
    UpsertRoom {
        room: Room,
    },
    UpsertThread {
        thread: Thread,
    },
    UpsertMessage {
        message: Message,
    },
    UpsertUser {
        user: User,
    },
    UpsertMember {
        member: RoomMember,
    },
    UpsertSession {
        session: Session,
    },
    UpsertRole {
        role: Role,
    },
    UpsertInvite {
        invite: Invite,
    },
    DeleteMessage {
        thread_id: ThreadId,
        message_id: MessageId,
    },
    DeleteMessageVersion {
        thread_id: ThreadId,
        message_id: MessageId,
        version_id: MessageVerId,
    },
    DeleteUser {
        id: UserId,
    },
    DeleteSession {
        id: SessionId,
    },
    DeleteRole {
        room_id: RoomId,
        role_id: RoleId,
    },
    DeleteMember {
        room_id: RoomId,
        user_id: UserId,
    },
    DeleteInvite {
        code: InviteCode,
    },
    Webhook {
        hook_id: Uuid,
        data: serde_json::Value,
    },
}

// /// messages specific to a user
// #[derive(Debug, PartialEq, Eq, ToSchema, Serialize, Deserialize)]
// #[serde(tag = "type")]
// enum MessageUser {}

// /// messages specific to a thread
// #[derive(Debug, PartialEq, Eq, ToSchema, Serialize, Deserialize)]
// #[serde(tag = "type")]
// enum MessageThread {}

// /// messages specific to a room
// #[derive(Debug, PartialEq, Eq, ToSchema, Serialize, Deserialize)]
// #[serde(tag = "type")]
// enum MessageRoom {}

type WsMessage = axum::extract::ws::Message;

impl From<MessageServer> for WsMessage {
    fn from(value: MessageServer) -> Self {
        WsMessage::text(
            serde_json::to_string(&value)
                .expect("servermessage should always be able to be serialized"),
        )
    }
}
